### PR TITLE
Network Mux Close

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

**Review by njd42 — APPROVED** [review]
This is all looking good. The logic of the "group" didn't jump out at me, but thats fine.

File: network-mux/src/Network/Mux.hs [hunk 1]
```diff
@@ -406,8 +410,17 @@ monitor tracer timeout jobpool egressQueue cmdQueue muxStatus =
...
+            size <- JobPool.readGroupSize jobpool MiniProtocolJob
+            case size of
+              0  | Just (MuxError MuxBearerClosed _) <- fromException e
+                -> writeTVar muxStatus MuxStopped
+                >> return True
+              _ -> writeTVar muxStatus (MuxFailed e)
+                >> return False
+          if r
+            then return ()
+            else throwIO e
```
Comment (karknu) [inline network-mux/src/Network/Mux.hs:423]: :+1: no more 'MuxError BearerClosed' questions!

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+--
+data CloseError = CloseTimeoutError
+  deriving (Show, Typeable)
+
+instance Exception CloseError
+
+
+data SuccessOrFailure = Success | Failure
+
+type Size = Int64
```
Comment (karknu) [inline network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:145]: Why is this `Int64` and not `Word64`?

**Review by karknu — APPROVED** [review]
LGTM.

File: network-mux/src/Network/Mux.hs (hunk 1 above)
Comment (coot) [reply network-mux/src/Network/Mux.hs:423]: but only in the `p2p` branch. Clean close needs to await for all mini-protocols to terminate and we don't have such a component in the current system.  It wouldn't be difficult to implement though.

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs (hunk 2 above)
Comment (coot) [reply network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:145]: The only reason is that lazy bytestring is using `Int64` for [`length`](https://hackage.haskell.org/package/bytestring-0.11.1.0/docs/Data-ByteString-Lazy.html#v:length).

**Comment by coot** [comment]
bors merge

**Pull request merged**